chore: add MAINTAIN support for postgres >= 17

sudo-dliberty-eiq opened the pull request:
## Summary

- Adds `MAINTAIN` privilege support for PostgreSQL >= 17 tables
- Makes privilege resolution version-aware so `ALL` expands correctly on each PG version
- Adds PG 17 to the CI test matrix

## Problem

The `MAINTAIN` privilege was added to the static `allowedPrivileges` map unconditionally. On PG 12-16, `GRANT ALL ON TABLE` returns 7 privileges (no MAINTAIN), but `resourcePrivilegesEqual()` expanded `ALL` to 8 privileges — causing all acceptance tests to fail.

## Approach

- Added `featureMaintainPrivilege` feature flag gated on `>=17.0.0`
- Replaced static privilege lookups with `allowedPrivilegesForObjectType(objectType, db)` which appends `MAINTAIN` only when the connected PG version supports it
- Threaded `*DBConnection` through `validatePrivileges`, `resourcePrivilegesEqual`, and the read-privilege functions so version detection works at all call sites
- When `db` is nil (e.g. unit tests), only base privileges are returned

## Test plan

- [x] `go test ./postgresql/... -run TestArePrivilegesEqual` passes (unit test with 7 base privileges)
- [ ] CI tests on PG 12-16 pass (MAINTAIN not included in ALL expansion)
- [ ] CI test on PG 17 passes (MAINTAIN included in ALL expansion)

sudo-dliberty-eiq committed:
chore: add MAINTAIN support for postgres >= 17

yakloinsteak commented:
I needed this fix as well, but though it now is able to apply `MAINTAIN`, if you run a plan after that, it seems that postgres has "expanded" this into individual permissions, so you're forever out of sync, oscillating between `MAINTAIN` and that set.

This is what happens. It either wants to do this or add the `MAINTAIN` priv.

```
~ privileges        = [
    - "MAINTAIN",
    + "DELETE",
    + "INSERT",
    + "REFERENCES",
    + "SELECT",
    + "TRIGGER",
    + "TRUNCATE",
    + "UPDATE",
  ]
```

I think the fix might need to account for this, but I haven't done a lot of testing. I'm going to replace all of our `MAINTAIN`s with this set and move on, though supporting this priv would be nice.

ooaklee commented:
Any update on this?

sudo-dliberty-eiq commented:
Just waiting on approval to merge by a reviewer with write access

mohammedelnabawy commented:
Any update on this?

lsowen commented:
@cyrilgdn would you be able to review this change?

logan-hcg commented:
I don't see the "expanded permissions" issue mentioned above when running against a postgres 17 server.

cyrilgdn committed:
Merge branch 'main' into maintain_permisison

chafouin commented:
@cyrilgdn I saw you made a release recently, would it be possible to review/merge this PR and make another release please ? Thanks 🙇

sudo-dliberty-eiq committed:
Merge branch 'main' into maintain_permisison

romanpanov993 commented:
@cyrilgdn please, review it

Peter4net commented:
Hi, we also need this, when can expect this be released? Thank you!

ahathaway commented:
@cyrilgdn 

What is happening with this test which didn't pass? the logs are too old to view at this point. This is pretty much a critical feature for anyone using this provider on a version of postgres after the introduction of MAINTAIN. What do we need to do in order to advance this? We are coming up on almost 1 year since this issue was filed, it would be greatly appreciated if we could get this PR merged in and released.

philip-harvey commented:
@cyrilgdn also weighing in, this is urgently needed to support PostgreSQL 17

sudo-dliberty-eiq committed:
fix: make MAINTAIN privilege version-aware (PG >= 17)

The MAINTAIN privilege is only available in PostgreSQL 17+. Previously it
was added to the static allowedPrivileges map unconditionally, which
broke ALL privilege expansion on PG 12-16 (8 vs 7 privileges mismatch).

This commit introduces a featureMaintainPrivilege feature flag gated on
>=17.0.0 and makes allowedPrivileges resolution version-aware via a new
allowedPrivilegesForObjectType helper. Also adds PG 17 to the CI matrix.

sudo-dliberty-eiq commented:
@cyrilgdn Gave this one a refresh, please review when you get the chance. Thank you!

randyhbh commented:
Hi guys, any update on this? I have been waiting for a long time.

bezalelbrener commented:
hey, waiting for this as well. i see its approved, waiting for merge @sudo-dliberty-eiq

cyrilgdn committed:
Merge branch 'main' into maintain_permisison

cyrilgdn reviewed (approved):
Thanks for your work 💪

imishchuk-tsgs commented:
A release with this fix would be nice, please.
